Source/ResponseSerialization.swift

```diff
@@ -798,3 +798,151 @@ extension DownloadRequest {
                  completionHandler: completionHandler)
     }
 }
+
+// MARK: - DataStreamRequest
+
+/// A type which can serialize incoming `Data`.
+public protocol DataStreamSerializer {
+    /// Type produced from the serialized `Data`.
+    associatedtype SerializedObject
+
+    /// Serializes incoming `Data` into a `SerializedObject` value.
+    ///
+    /// - Parameter data: `Data` to be serialized.
+    ///
+    /// - Throws: Any error produced during serialization.
+    func serialize(_ data: Data) throws -> SerializedObject
+}
+
+/// `DataStreamSerializer` which uses the provided `DataPreprocessor` and `DataDecoder` to serialize the incoming `Data`.
+public struct DecodableStreamSerializer<T: Decodable>: DataStreamSerializer {
+    public let decoder: DataDecoder
+    public let dataPreprocessor: DataPreprocessor
+
+    /// Creates an instance with the provided `DataDecoder` and `DataPreprocessor`.
+    /// - Parameters:
+    ///   - decoder: `        DataDecoder` used to decode incoming `Data`.
+    ///   - dataPreprocessor: `DataPreprocessor` used to process incoming `Data` before it's passed through the `decoder`.
+    public init(decoder: DataDecoder = JSONDecoder(), dataPreprocessor: DataPreprocessor = PassthroughPreprocessor()) {
+        self.decoder = decoder
+        self.dataPreprocessor = dataPreprocessor
+    }
+
+    public func serialize(_ data: Data) throws -> T {
+        let processedData = try dataPreprocessor.preprocess(data)
+        do {
+            return try decoder.decode(T.self, from: processedData)
```

Comment:
How do you know you have enough data to actually run the decode pass safely here? Not really sure how this actually works.

Reply:
We don't. Unlike web sockets, this sort of streaming is completely unstructured and at the mercy of how `URLSession` receives the `Data`. I'm pretty sure `Data` which fits entirely within a single MTU frame will come through unbroken, and I've confirmed that to be the case with some streaming servers that stream small JSON payloads. But larger payloads likely won't work, but I didn't want to build a whole custom accumulator that would then feed into serializers, as the complexity of that would be very high. Plus, that scenario would be better covered by a web socket, which has automatic framing built in to make that sort of thing much easier as I understand it (though I'm working to confirm that behavior). 

Ultimately, if users find their `Data` is too big for the automatic handling, I think they could accumulate the `Data` themselves and trigger parsing using the serializers manually.

Reply:
Okay cool. That is what I "assumed" it was built for, but wanted to confirm. So this is much more of a "use at your own risk" type of feature.